plugin/color_picker/__init__.py

```diff
@@ -0,0 +1,239 @@
+            ("Flags", ctypes.c_uint32),
+            ("lCustData", ctypes.c_void_p),
+            ("lpfnHook", ctypes.c_void_p),
+            ("lpTemplateName", ctypes.c_wchar_p)]
+
+    CC_SOLIDCOLOR = 0x80
+    CC_RGBINIT = 0x01
+    CC_FULLOPEN = 0x02
+    ChooseColorW = ctypes.windll.Comdlg32.ChooseColorW
+    ChooseColorW.argtypes = [ctypes.POINTER(CHOOSECOLOR)]
+    ChooseColorW.restype = ctypes.c_int32
+
+
+    class WindowsColorPicker(ColorPickerPlugin):
+        def pick(self, on_pick: OnPickCallback, preselect_color: Optional[Color] = None) -> None:
+            self.close()
+            default_color = (255 << 16) | (255 << 8) | (255)
+            if preselect_color:
+                default_color = (round(255*preselect_color['blue']) << 16) | (round(255*preselect_color['green']) << 8) | round(255*preselect_color['red'])
+            cc = CHOOSECOLOR()
+            ctypes.memset(ctypes.byref(cc), 0, ctypes.sizeof(cc))
+            cc.lStructSize = ctypes.sizeof(cc)
+            cc.hwndOwner = sublime.active_window().hwnd()
+            CustomColors = ctypes.c_uint32 * 16
+            cc.lpCustColors = CustomColors() # uses 0 (black) for all 16 predefined custom colors
+            cc.rgbResult = ctypes.c_uint32(default_color)
+            cc.Flags = CC_SOLIDCOLOR | CC_FULLOPEN | CC_RGBINIT
+
+            # ST window will become unresponsive until color picker dialog is closed
+            output = ChooseColorW(ctypes.byref(cc))
+
+            if output == 1: # user clicked OK
+                on_pick(self.normalize_color(cc.rgbResult))
+            else:
+                on_pick(None)
+
+        def normalize_color(self, bgr_color: Any) -> Optional[Color]:
+
+            def bgr2color(bgr) -> Color:
+                # 0x00BBGGRR
+                byte_table = list(["{0:02X}".format(b) for b in range(256)])
+                b_hex = byte_table[(bgr >> 16) & 0xff]
+                g_hex = byte_table[(bgr >> 8) & 0xff]
+                r_hex = byte_table[(bgr) & 0xff]
+
+                r = int(r_hex, 16) / 255
+                g = int(g_hex, 16) / 255
+                b = int(b_hex, 16) / 255
+                return {
+                    "red": r,
+                    "green": g,
+                    "blue": b,
+                    "alpha": 1  # Windows color picker doesn't support alpha, so fallback to 1
+                }
+
+            if bgr_color:
+                return bgr2color(bgr_color)
+            return None
+
+        def close(self) -> None:
+            pass # on windows, the color picker will block until a color is choosen 
+
+
+class OsxColorPicker(ColorPickerPlugin):
+    process = None  # type: Optional[subprocess.Popen]
+    script = '''
+    function run(argv) {{
+        var app = Application.currentApplication();
+        app.includeStandardAdditions = true;
+        var color = app.chooseColor({{defaultColor: [{red}, {green}, {blue}]}});
+        console.log(JSON.stringify(color));
+    }}
+    '''
```

Comment:
I suppose that updating the view color is not currently supported?

Reply:
Yes, that is due to this point:

> ❌ Mac - When we chose a color, OSX output contains the following warning (and a color at the end):
2022-09-14 10:13:38.606 osascript[66572:9484254] It's not legal to call -layoutSubtreeIfNeeded on a view which is already being laid out. If you are implementing the view's -layout method, you can call -[super layout] instead. Break on void _NSDetectedLayoutRecursion(void) to debug. This will be logged only once. This may break in the future.
[0.16188296675682068,1,0.7679865956306458]

I could write a regex to extract the error, but I didn't want to fix that, maybe someone has a better way to fix that error.